app/core/src/main/java/com/fsck/k9/controller/PermanentDeleteChecker.kt

```diff
@@ -0,0 +1,83 @@
+package com.fsck.k9.controller
+
+import androidx.annotation.WorkerThread
+import com.fsck.k9.Account
+import com.fsck.k9.Account.DeletePolicy
+import com.fsck.k9.preferences.AccountManager
+
+/**
+ * Checks whether deleting a list of messages in the app will permanently delete none, some, or all of them.
+ *
+ * Deleting a message permanently means both the local copy and the server copy will be deleted.
+ */
+class PermanentDeleteChecker internal constructor(
+    private val accountManager: AccountManager,
+    private val localFolderChecker: LocalFolderChecker,
+    private val deleteOperationDecider: DeleteOperationDecider,
+) {
+    /**
+     * Check whether deleting messages will delete some of them permanently.
+     *
+     * Note: This method can perform disk I/O. Don't call it from the main thread!
+     */
+    @WorkerThread
+    fun checkPermanentDelete(messageReferences: List<MessageReference>): PermanentDeleteResult {
+        val permanentDeleteCount = messageReferences
+            .groupBy { it.accountUuid }
+            .mapKeysNotNull { accountUuid ->
+                accountManager.getAccount(accountUuid)
+            }
+            .map { (account, messageReferences) ->
+                getPermanentDeleteCount(account, messageReferences)
+            }
+            .sum()
+
+        return when (permanentDeleteCount) {
+            0 -> PermanentDeleteResult.None
+            messageReferences.size -> PermanentDeleteResult.All
+            else -> PermanentDeleteResult.Some(permanentDeleteCount)
+        }
+    }
+
+    private fun getPermanentDeleteCount(account: Account, messageReferences: List<MessageReference>): Int {
+        return messageReferences
+            .groupBy { it.folderId }
+            .map { (folderId, messageReferences) ->
```

Comment:
This threats all `MessageReference` related to a specific `folderId` in the same way. Is it possible for certain messages to be marked as "do not delete"?